implementations/rust/ockam/ockam_entity/src/profile.rs

```diff
@@ -0,0 +1,135 @@
+/// Profile is an abstraction responsible for keeping, verifying and modifying
+/// user's data (mainly - public keys). It is used to create new keys, rotate and revoke them.
+/// Public keys together with metadata will be organised into events chain, corresponding
+/// secret keys will be saved into the given Vault implementation. Events chain and corresponding
+/// secret keys are what fully determines Profile.
+///
+///
+/// # Examples
+///
+/// Create a [`Profile`]. Add and rotate keys.
+/// TODO
+///
+/// Authentication using [`Profile`]. In following example Bob authenticates Alice.
+/// TODO
+///
+/// Update [`Profile`] and send changes to other parties. In following example Alice rotates
+/// her key and sends corresponding [`Profile`] changes to Bob.
+/// TODO
+///
+use crate::{
+    Changes, Contact, Entity, Handle, Identity, ProfileChangeEvent, ProfileIdentifier, Proof,
+};
+use ockam_core::Result;
+use ockam_vault::{PublicKey, Secret};
+
+pub struct Profile {
+    id: ProfileIdentifier,
+    handle: Handle,
+}
+
+impl From<&mut Profile> for Entity {
+    fn from(p: &mut Profile) -> Entity {
+        Entity::new(p.handle.clone(), p.id.clone())
+    }
+}
+
+impl Profile {
+    pub fn new<I: Into<ProfileIdentifier>>(id: I, handle: Handle) -> Self {
+        let id = id.into();
+        Profile { id, handle }
+    }
+
+    pub fn entity(&mut self) -> Entity {
+        Entity::from(self)
+    }
+}
+
+impl Profile {
+    /// Sha256 of that value is used as previous event id for first event in a [`Profile`]
+    pub const NO_EVENT: &'static [u8] = "OCKAM_NO_EVENT".as_bytes();
+    /// Label for [`Profile`] update key
+    pub const PROFILE_UPDATE: &'static str = "OCKAM_PUK";
+    /// Label for key used to issue credentials
+    pub const CREDENTIALS_ISSUE: &'static str = "OCKAM_CIK";
+    /// Current version of change structure
+    pub const CURRENT_CHANGE_VERSION: u8 = 1;
+}
+
+impl Identity for Profile {
+    fn identifier(&mut self) -> Result<ProfileIdentifier> {
+        self.entity().identifier()
```

Comment:
Don't really understand why would Profile delegate to Entity and not vice versa

Reply:
Let's go over the full architecture with Mrinal soon

Reply:
Discussion is larger than this PR since we wanna merge it in the morning. Noting to talk about it later